crates/codegen/parser/runtime/src/support/precedence_helper.rs

```diff
@@ -63,172 +63,162 @@ impl PrecedenceHelper {
+        while elements.len() > 1 {
+            // 1. Find the next, highest priority reducable operator
+            // The elements are guaranteed to match the following:
+            // `prefixop* expr postfixop* ( binaryop prefixop* expr postfixop* )*`
+
+            match &elements[i..] {
+                [Expression { .. }, Postfix { .. }, ..] => {
+                    i += 1;
+                    continue;
+                }
+
+                [Expression { .. }, Binary { .. }, ..] => {
+                    i += 1;
+                    continue;
+                }
+
+                [Prefix { right, .. }, Expression { .. }, Binary { left, .. }, ..] => {
+                    if right <= left {
+                        i += 2;
+                        continue;
+                    } else {
+                        /* Reduce */
+                    }
+                }
+
+                [Prefix { right, .. }, Expression { .. }, Postfix { left, .. }, ..] => {
+                    if right <= left {
+                        i += 2;
+                        continue;
+                    } else {
+                        /* Reduce */
+                    }
+                }
+
+                [Prefix { .. }, Expression { .. }] => { /* Reduce */ }
+
+                [Prefix { .. }, Prefix { .. }, ..] => {
+                    i += 1;
+                    continue;
+                }
+
+                [Binary { right, .. }, Expression { .. }, Binary { left, .. }, ..] => {
+                    if right <= left {
+                        i += 2;
+                        continue;
+                    } else {
+                        /* Reduce */
+                    }
+                }
+
+                [Binary { right, .. }, Expression { .. }, Postfix { left, .. }, ..] => {
+                    if right <= left {
+                        i += 2;
+                        continue;
+                    } else {
+                        /* Reduce */
                     }
+                }
+
+                [Binary { .. }, Prefix { .. }, ..] => {
+                    i += 1;
+                    continue;
+                }
+
+                [Binary { .. }, Expression { .. }] => { /* Reduce */ }
+
+                [Postfix { .. }, ..] => { /* Reduce */ }
 
-                    i = i.saturating_sub(1);
+                _ => {
+                    unreachable!(
+                        "Unmatched precedence pattern at index {} in: {:#?}",
+                        i, elements
+                    )
                 }
+            };
+
+            // 2. Reduce the operator and it's child expressions to a new expression
+
+            let make_expression = |left: Option<PrattElement>,
+                                   kind: RuleKind,
+                                   nodes: Vec<cst::NamedNode>,
+                                   right: Option<PrattElement>| {
+                let &[left_name, right_name] = match (&left, &right) {
```

Comment:
Thoughts about using `None` instead of `!`? this way, we can always `.expect()` it and fail if something is wrong.

Reply:
I want to avoid expects/unwraps if it's possible; I've pushed 95fd624ec that simplifies this logic.